src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Unix.cs

```diff
@@ -0,0 +1,604 @@
+            }
+
+            internal ReadOnlySpan<SafeBagAsn> GetKeysSpan()
+            {
+                return new ReadOnlySpan<SafeBagAsn>(_keyBags, 0, _keyCount);
+            }
+        }
+
+        private struct CertAndKey
+        {
+            internal ICertificatePalCore? Cert;
+            internal AsymmetricAlgorithm? Key;
+
+            internal void Dispose()
+            {
+                Cert?.Dispose();
+                Key?.Dispose();
+            }
+        }
+
+        private struct CertKeyMatcher
+        {
+            private CertAndKey[] _certAndKeys;
+            private int _certCount;
+            private AsymmetricAlgorithm?[] _keys;
+            private RentedSubjectPublicKeyInfo[] _rentedSpki;
+            private int _keyCount;
+
+            internal void LoadCerts(ref BagState bagState)
+            {
+                if (bagState.CertCount == 0)
+                {
+                    return;
+                }
+
+                _certAndKeys = ArrayPool<CertAndKey>.Shared.Rent(bagState.CertCount);
+
+                foreach (SafeBagAsn safeBag in bagState.GetCertsSpan())
+                {
+                    Debug.Assert(safeBag.BagId == Oids.Pkcs12CertBag);
+
+                    CertBagAsn certBag = CertBagAsn.Decode(safeBag.BagValue, AsnEncodingRules.BER);
+
+                    // Non-X.509 cert-type bags should have already been removed.
+                    Debug.Assert(certBag.CertId == Oids.Pkcs12X509CertBagType);
+                    ReadOnlyMemory<byte> certData = Helpers.DecodeOctetStringAsMemory(certBag.CertValue);
+
+                    ICertificatePalCore pal = LoadX509Der(certData);
+                    Debug.Assert(pal is not null);
+
+                    _certAndKeys[_certCount].Cert = pal;
+                    _certCount++;
+                }
+            }
+
+            internal void LoadKeys(ref BagState bagState)
+            {
+                if (bagState.KeyCount == 0)
+                {
+                    return;
+                }
+
+                _keys = ArrayPool<AsymmetricAlgorithm?>.Shared.Rent(bagState.KeyCount);
+
+                foreach (SafeBagAsn safeBag in bagState.GetKeysSpan())
+                {
+                    AsymmetricAlgorithm? key = null;
+
+                    try
+                    {
+                        if (safeBag.BagId == Oids.Pkcs12KeyBag)
+                        {
+                            PrivateKeyInfoAsn privateKeyInfo =
+                                PrivateKeyInfoAsn.Decode(safeBag.BagValue, AsnEncodingRules.BER);
+
+                            key = CreateKey(privateKeyInfo.PrivateKeyAlgorithm.Algorithm);
+
+                            if (key is not null)
+                            {
+                                // No need to check bytesRead since it was verified
+                                // in PrivateKeyInfoAsn.Decode
+                                key.ImportPkcs8PrivateKey(safeBag.BagValue.Span, out int _);
```

Comment:
On macOS, this will throw a PNSE for unsupported EC curves (like brainpool) and some tests are expecting a CryptographicException to be thrown. Maybe something like this

```suggestion
                                try
                                {
                                    key.ImportPkcs8PrivateKey(safeBag.BagValue.Span, out int _);
                                }
                                catch (PlatformNotSupportedException pe)
                                {
                                    throw new CryptographicException(SR.Cryptography_NotValidPrivateKey, pe);
                                }
```

Reply:
You will need the same fix on line 393.

Reply:
I'm not sure that the second instance is important (since it should have already hit the first) one; but I went ahead and made a helper function and routed them both through it.